### ISSUE TITLE
gateway packages for ubuntu require libffi6 when libffi7 is packaged in Ubuntu 20.04 and 20.10

### ISSUE THREAD

**Issue by chas-iot** [WebThingsIO/gateway#2681]
Do you need any further info?
Is there a work-around?

**Comment by chas-iot** [comment]
@mrstegeman is there a set of instructions anywhere for me to build a new DEB so I can test this? I think that there are more issues in the package but I've no clue how to even start.

**Comment by mrstegeman** [comment]
Clone the gateway, then:
```
cd deb
## edit build-deb-manual.sh with your desired package
./build-deb-manual.sh
```

**Comment by chas-iot** [comment]
I have it building :)
It builds from tag v1.0.0 not head :(
I'll see what I can work out but advice will be appreciated

**Comment by chas-iot** [comment]
Maybe that's the tag for gateway-addon-python, not the gateway - never mind, the build is continuing - I'll see whether I get results at the end

**Comment by mrstegeman** [comment]
I know about at least one other issue, which I don’t have a succinct fix for yet: https://github.com/WebThingsIO/gateway-aur/issues/4

If you can find a fix for that, I would be incredibly grateful!

**Comment by chas-iot** [comment]
success! - at least first step
````
charles@s4:~/gateway/deb$ sudo apt install ./webthings-gateway-ubuntu-focal-amd64.deb
Reading package lists... Done
Building dependency tree
Reading state information... Done
Note, selecting 'webthings-gateway' instead of './webthings-gateway-ubuntu-focal-amd64.deb'
The following additional packages will be installed:
  alsa-utils libatopology2 libfftw3-single3 python-six
Suggested packages:
  dialog libfftw3-bin libfftw3-dev
The following NEW packages will be installed:
  alsa-utils libatopology2 libfftw3-single3 python-six webthings-gateway
0 upgraded, 5 newly installed, 0 to remove and 0 not upgraded.
Need to get 1837 kB/15.8 MB of archives.
After this operation, 73.5 MB of additional disk space will be used.
Do you want to continue? [Y/n]
Get:1 /home/charles/gateway/deb/webthings-gateway-ubuntu-focal-amd64.deb webthings-gateway amd64 1.0.0-1 [14.0 MB]
Get:2 http://sg.archive.ubuntu.com/ubuntu focal-updates/main amd64 libatopology2 amd64 1.2.2-2.1ubuntu2.2 [47.3 kB]
Get:3 http://sg.archive.ubuntu.com/ubuntu focal/main amd64 libfftw3-single3 amd64 3.3.8-2ubuntu1 [756 kB]
Get:4 http://sg.archive.ubuntu.com/ubuntu focal-updates/main amd64 alsa-utils amd64 1.2.2-1ubuntu2 [1021 kB]
Get:5 http://sg.archive.ubuntu.com/ubuntu focal/universe amd64 python-six all 1.14.0-2 [12.0 kB]
Fetched 1837 kB in 8s (239 kB/s)
Selecting previously unselected package libatopology2:amd64.
(Reading database ... 106367 files and directories currently installed.)
Preparing to unpack .../libatopology2_1.2.2-2.1ubuntu2.2_amd64.deb ...
Unpacking libatopology2:amd64 (1.2.2-2.1ubuntu2.2) ...
Selecting previously unselected package libfftw3-single3:amd64.
Preparing to unpack .../libfftw3-single3_3.3.8-2ubuntu1_amd64.deb ...
Unpacking libfftw3-single3:amd64 (3.3.8-2ubuntu1) ...
Selecting previously unselected package alsa-utils.
Preparing to unpack .../alsa-utils_1.2.2-1ubuntu2_amd64.deb ...
Unpacking alsa-utils (1.2.2-1ubuntu2) ...
Selecting previously unselected package python-six.
Preparing to unpack .../python-six_1.14.0-2_all.deb ...
Unpacking python-six (1.14.0-2) ...
Selecting previously unselected package webthings-gateway.
Preparing to unpack .../webthings-gateway-ubuntu-focal-amd64.deb ...
Unpacking webthings-gateway (1.0.0-1) ...
Setting up libfftw3-single3:amd64 (3.3.8-2ubuntu1) ...
Setting up python-six (1.14.0-2) ...
Setting up libatopology2:amd64 (1.2.2-2.1ubuntu2.2) ...
Setting up webthings-gateway (1.0.0-1) ...
usermod: group 'gpio' does not exist
Setting up alsa-utils (1.2.2-1ubuntu2) ...
alsa-restore.service is a disabled or a static unit, not starting it.
alsa-state.service is a disabled or a static unit, not starting it.
Processing triggers for systemd (245.4-4ubuntu3.3) ...
Processing triggers for man-db (2.9.1-1) ...
Processing triggers for libc-bin (2.31-0ubuntu9.1) ...

**Comment by chas-iot** [comment]
and logged in to running gateway on Ubuntu 20.04

**Comment by mrstegeman** [comment]
So I guess the launcher is working ok for you? I wonder if it’s only an issue on 20.10.

**Comment by chas-iot** [comment]
All seems OK. I am using 20.04 as it is LTS, better as a stable platform for development. I'm about to reboot, let's see if it comes up after restart

**Comment by chas-iot** [comment]
Yes, restarted correctly after reboot

**Comment by mrstegeman** [comment]
Well, that’s good. Guess I’ll have to figure the issue out myself then. 😒

**Comment by mrstegeman** [comment]
Also, since you built from master, can you confirm that the update page in Settings tells you that updates are not supported?

**Comment by chas-iot** [comment]
confirmed

**Comment by mrstegeman** [comment]
Fantastic, thanks!

**Comment by chas-iot** [comment]
for full proof that the deb is OK, I will create a temporary VM from scratch and install there. If I do that I can do it for both 20.04 and 20.10

**Comment by mrstegeman** [comment]
That would be helpful, for 20.10. Don’t forget you’ll have to build a separate package for that.

**Comment by chas-iot** [comment]
needs work - I'll start looking
This is on a fresh install of 20.04 from ISO image, with no apt updates
````
apt install ./webthings-gateway-ubuntu-focal-amd64.deb
Reading package lists... Done
Building dependency tree
Reading state information... Done
Note, selecting 'webthings-gateway' instead of './webthings-gateway-ubuntu-focal-amd64.deb'
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 webthings-gateway : Depends: nodejs (>= 10.0.0) but it is not installable
                     Depends: nodejs (< 11.0.0~~) but it is not installable
                     Depends: python-six but it is not installable
                     Depends: python3-pip but it is not installable
                     Recommends: alsa-utils but it is not going to be installed
                     Recommends: arping but it is not installable
                     Recommends: avahi-daemon but it is not going to be installed
                     Recommends: ffmpeg but it is not installable
                     Recommends: mosquitto but it is not installable
                     Recommends: sqlite3 but it is not going to be installed
E: Unable to correct problems, you have held broken packages.

**Comment by mrstegeman** [comment]
Make sure you run `apt update` just to pull down package lists.

**Comment by chas-iot** [comment]
:D

**Comment by chas-iot** [comment]
It's installing after `apt update; apt full-upgrade`

**Comment by chas-iot** [comment]
And complete, with no issues on 20.04 :)
I'm building the deb for 20.10 and creating a new VM - will post an update later

**Comment by chas-iot** [comment]
the results for 20.10 on ARM64 - not so good :(

````
apt install ./webthings-gateway-ubuntu-groovy-amd64.deb
Reading package lists... Done
Building dependency tree
Reading state information... Done
Note, selecting 'webthings-gateway' instead of './webthings-gateway-ubuntu-groovy-amd64.deb'
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 webthings-gateway : Depends: libffi8 but it is not installable
                     Recommends: alsa-utils but it is not going to be installed
                     Recommends: arping but it is not going to be installed
                     Recommends: avahi-daemon but it is not going to be installed
                     Recommends: ffmpeg but it is not going to be installed
                     Recommends: mosquitto but it is not going to be installed
                     Recommends: sqlite3 but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
````

````
apt list 'libffi*'
Listing... Done
libffi-checklib-perl/groovy 0.27-1 all
libffi-dev/groovy 3.4~20200819gead65ca871-0ubuntu3 amd64
libffi-platypus-perl/groovy 1.31-1build2 amd64
libffi8ubuntu1/groovy,now 3.4~20200819gead65ca871-0ubuntu3 amd64 [installed,automatic]
libffindex0-dev/groovy 0.9.9.9-3 amd64
libffindex0/groovy 0.9.9.9-3 amd64

**Comment by mrstegeman** [comment]
Ubuntu always has to make things difficult. I just pushed up bf5f69b89acdaeaaf4c7017c21b0c2f6307f82f0. Could you give that a shot?

**Comment by chas-iot** [comment]
It's installing now
````
apt install ./webthings-gateway-ubuntu-groovy-amd64.deb
Reading package lists... Done
Building dependency tree
Reading state information... Done
Note, selecting 'webthings-gateway' instead of './webthings-gateway-ubuntu-groovy-amd64.deb'
The following additional packages will be installed:

**Comment by chas-iot** [comment]
And it's running and I can connect

````
systemctl status webthings-gateway.service
● webthings-gateway.service - WebThings Gateway
     Loaded: loaded (/lib/systemd/system/webthings-gateway.service; enabled; vendor preset: enabled)
     Active: active (running) since Sun 2020-12-13 02:18:59 UTC; 2min 26s ago
   Main PID: 16550 (webthings-gatew)
      Tasks: 12 (limit: 986)
     Memory: 86.0M
     CGroup: /system.slice/webthings-gateway.service
             ├─16550 /bin/bash /usr/bin/webthings-gateway
             └─16553 node build/gateway.js

Dec 13 02:18:59 s5 systemd[1]: Started WebThings Gateway.
Dec 13 02:19:11 s5 webthings-gateway[16553]: Creating database: /etc/webthings-gateway/config/db.sqlite3
Dec 13 02:19:23 s5 webthings-gateway[16553]: 2020-12-13 02:19:23.771 INFO   : Creating database: /etc/webthings-gateway/log/logs.sqlite3
Dec 13 02:19:26 s5 webthings-gateway[16553]: 2020-12-13 02:19:26.050 INFO   : HTTP server listening on port 8080
Dec 13 02:19:31 s5 webthings-gateway[16553]: 2020-12-13 02:19:31.722 DEBUG  : Ignoring https://raw.githubusercontent.com/WebThingsIO/gateway-addon-ipc-schema/master/schema.json because it has no messageType
Dec 13 02:19:32 s5 webthings-gateway[16553]: 2020-12-13 02:19:32.766 DEBUG  : Ignoring https://raw.githubusercontent.com/WebThingsIO/gateway-addon-ipc-schema/master/messages/definitions.json because it has no messageType
Dec 13 02:19:44 s5 webthings-gateway[16553]: 2020-12-13 02:19:44.539 INFO   : Checking for add-on updates...
Dec 13 02:19:45 s5 webthings-gateway[16553]: 2020-12-13 02:19:45.856 INFO   : Finished updating add-ons
````
![image](https://user-images.githubusercontent.com/65324897/102001461-7e7ee480-3d2d-11eb-91aa-0fd2f41a0cbd.png)

**Comment by mrstegeman** [comment]
Still no issues with the launcher, then?

**Comment by chas-iot** [comment]
After the install completed, the gateway was running with no further intervention from me.

**Comment by mrstegeman** [comment]
👍🏻

**Comment by chas-iot** [comment]
> the results for 20.10 on ARM64 - not so good :(

should read AMD64 - not ARM

But I think this issue can be well and truly closed now

**Comment by chas-iot** [comment]
> I know about at least one other issue, which I don’t have a succinct fix for yet: [WebThingsIO/gateway-aur#4](https://github.com/WebThingsIO/gateway-aur/issues/4)
> 
> If you can find a fix for that, I would be incredibly grateful!

I think that this is unrelated - I saw no issue like it

**Comment by mrstegeman** [comment]
> > I know about at least one other issue, which I don’t have a succinct fix for yet: [WebThingsIO/gateway-aur#4](https://github.com/WebThingsIO/gateway-aur/issues/4)
> > If you can find a fix for that, I would be incredibly grateful!
> 
> I think that this is unrelated - I saw no issue like it

Ok, that’s good to hear.
